File: backend/futures-matching/src/main/java/com/futures/matching/model/Order.java
package com.futures.matching.model;

import lombok.Builder;
import lombok.Data;

/**
 * 订单实体 — 不可变对象
 */
@Data
@Builder
public class Order {
    String orderId;
    String userId;
    String symbol;
    Direction direction;   // BUY / SELL
    OrderType type;        // LIMIT / MARKET / STOP / TAKE_PROFIT / FOK / IOC
    long price;            // 单位：最小变动价位整数
    int volume;            // 手数
    int filledVolume;      // 已成交量
    long timestamp;        // 创建时间戳（纳秒）

    public boolean isBuy() { return direction == Direction.BUY; }
    public boolean isSell() { return direction == Direction.SELL; }
    public boolean isLimit() { return type == OrderType.LIMIT; }
    public boolean isMarket() { return type == OrderType.MARKET; }

    public int remainingVolume() { return volume - filledVolume; }

    public enum Direction { BUY, SELL }
    public enum OrderType { LIMIT, MARKET, STOP, TAKE_PROFIT, FOK, IOC }
}

File: core/model/src/main/java/com/hackfuture/core/model/Order.kt
package com.hackfuture.core.model

import kotlinx.serialization.Serializable

/**
 * 交易订单
 */
@Serializable
data class Order(
    val id: String,
    val symbol: String,
    val type: OrderType,
    val side: OrderSide,
    val status: OrderStatus,
    val price: Double? = null,
    val stopPrice: Double? = null,
    val quantity: Double,
    val filledQuantity: Double = 0.0,
    val averageFilledPrice: Double? = null,
    val totalAmount: Double,
    val fee: Double = 0.0,
    val createdAt: Long,
    val updatedAt: Long,
) {
    val remainingQuantity: Double
        get() = quantity - filledQuantity

    val isFullyFilled: Boolean
        get() = status == OrderStatus.FILLED

    val isActive: Boolean
        get() = status == OrderStatus.PENDING || status == OrderStatus.PARTIALLY_FILLED
}

@Serializable
enum class OrderType {
    MARKET,
    LIMIT,
    STOP,
    STOP_LIMIT,
}

@Serializable
enum class OrderSide {
    BUY,
    SELL,
}

@Serializable
enum class OrderStatus {
    PENDING,
    PARTIALLY_FILLED,
    FILLED,
    CANCELLED,
    REJECTED,
    EXPIRED,
}

/**
 * 下单请求
 */
@Serializable
data class PlaceOrderRequest(
    val symbol: String,
    val type: OrderType,
    val side: OrderSide,
    val price: Double? = null,
    val stopPrice: Double? = null,
    val quantity: Double,
)

/**
 * 撤单请求
 */
@Serializable
data class CancelOrderRequest(
    val orderId: String,
    val symbol: String,
)

@Serializable
data class OrderHistoryItem(
    val orderId: String,
    val symbol: String,
    val side: OrderSide,
    val price: Double,
    val quantity: Double,
    val totalAmount: Double,
    val fee: Double,
    val executedAt: Long,
)
